databricks/koalas/tests/test_dataframe.py

```diff
@@ -368,6 +368,64 @@ def test_rename_columns(self):
         self.assert_eq(kdf._internal.data_columns, ["('A', '0')", "('B', 1)"])
         self.assert_eq(kdf._internal.spark_df.columns, ["('A', '0')", "('B', 1)"])
 
+    def test_rename_dataframe(self):
+        kdf1 = ks.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
+        result_kdf = kdf1.rename(columns={"A": "a", "B": "b"})
+        self.assert_eq(result_kdf.columns, pd.Index(['a', 'b']))
+
+        result_kdf = kdf1.rename(index={1: 10, 2: 20})
+        self.assert_eq(result_kdf.index, pd.Index([0, 10, 20]))
+
+        def str_lower(s) -> str:
+            return str.lower(s)
+
+        result_kdf = kdf1.rename(str_lower, axis='columns')
+        self.assert_eq(result_kdf.columns, pd.Index(['a', 'b']))
+
+        def mul10(x) -> int:
+            return x * 10
+
+        result_kdf = kdf1.rename(mul10, axis='index')
+        self.assert_eq(result_kdf.index, pd.Index([0, 10, 20]))
+
+        idx = pd.MultiIndex.from_tuples([('X', 'A'), ('X', 'B'), ('Y', 'C'), ('Y', 'D')])
+        kdf2 = ks.DataFrame([[1, 2, 3, 4], [5, 6, 7, 8]], columns=idx)
+
+        result_kdf = kdf2.rename(columns=str_lower)
+        self.assert_eq(result_kdf.columns,
+                       pd.MultiIndex.from_tuples([('x', 'a'), ('x', 'b'), ('y', 'c'), ('y', 'd')]))
+
+        result_kdf = kdf2.rename(columns=str_lower, level=0)
+        self.assert_eq(result_kdf.columns,
+                       pd.MultiIndex.from_tuples([('x', 'A'), ('x', 'B'), ('y', 'C'), ('y', 'D')]))
+
+        result_kdf = kdf2.rename(columns=str_lower, level=1)
+        self.assert_eq(result_kdf.columns,
+                       pd.MultiIndex.from_tuples([('X', 'a'), ('X', 'b'), ('Y', 'c'), ('Y', 'd')]))
+
+        kdf3 = ks.DataFrame([[1, 2], [3, 4], [5, 6], [7, 8]], index=idx, columns=list('ab'))
+
+        result_kdf = kdf3.rename(index=str_lower)
+        self.assert_eq(result_kdf.index,
+                       pd.Index([{'__index_level_0__': 'x', '__index_level_1__': 'a'},
```

Comment:
Oh wait, I thought it becomes a dictionary only against Spark 3 which is not released yet. Does this return a dictionary (due to Arrow optimization) in Spark 2.4 and 2.3 too.

Reply:
Actually returning `pd.Index` from `MultiIndex.to_pandas()` is a bug. It should be `pd.MultiIndex`.

Reply:
Yea but another problem was that it contained dictionaries (not tuples).. 

Given:

```python
>>> import pandas as pd
>>>
>>> pd.Index([('x', 'a'), ('x', 'b'), ('y', 'c'), ('y', 'd')], dtype='object')
MultiIndex(levels=[['x', 'y'], ['a', 'b', 'c', 'd']],
           codes=[[0, 0, 1, 1], [0, 1, 2, 3]])
```

I thought it might not be a big deal from a rough look. Will take a closer look tonight or tomorrow as planned.

Reply:
I think we need to override `MultiIndex.to_pandas()` to return `pd.MultiIndex`.

Reply:
We could fix this in separate PR.

Reply:
Yea, I think it should be a separate PR.

Reply:
Can we fix that issue first then?

Reply:
Let me fix it.

Reply:
Submitted a PR #832.